### PR TITLE
fix(rust,python): clean up `dt` namespace, fix `mean`/`median` return dtypes, impl `std` for `date`/`datetime`

### LINKED ISSUE #10232: Median on duration Series returns a datetime rather than a timedelta
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import timedelta
import polars as pl
s = pl.Series("s", [timedelta(0), timedelta(1)])
s.describe()
```

Output:
```
shape: (5, 2)
┌────────────┬─────────────────────┐
│ statistic  ┆ value               │
│ ---        ┆ ---                 │
│ str        ┆ str                 │
╞════════════╪═════════════════════╡
│ count      ┆ 2                   │
│ null_count ┆ 0                   │
│ min        ┆ 0:00:00             │
│ max        ┆ 1 day, 0:00:00      │
│ median     ┆ 1970-01-01 12:00:00 │
└────────────┴─────────────────────┘
```

### Issue description

The median shows up as a datetime in .describe() of a Series.  But in a dataframe, the result is correctly a timedelta/duration.

### Expected behavior

Expected:
```
shape: (5, 2)
┌────────────┬─────────────────────┐
│ statistic  ┆ value               │
│ ---        ┆ ---                 │
│ str        ┆ str                 │
╞════════════╪═════════════════════╡
│ count      ┆ 2                   │
│ null_count ┆ 0                   │
│ min        ┆ 0:00:00             │
│ max        ┆ 1 day, 0:00:00      │
│ median     ┆ 12:00:00            │
└────────────┴─────────────────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.11
Index type:          UInt32
Platform:            Linux-3.10.0-1160.92.1.el7.x86_64-x86_64-with-glibc2.17
Python:              3.9.17 (main, Jun  8 2023, 11:31:03) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               <not installed>
pandas:              <not installed>
pyarrow:             <not installed>
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
**This is a breaking change**

Resolves #10232, #12870, and #12883. @MarcoGorelli two notes to consider:

1. The breaking change here is that the `mean` and `median` functions for `pl.Date` now return a `datetime.datetime` instead of a `datetime.date`. This seemed much more logical if we consider dates corresponding to integers and datetimes corresponding to floats, since the mean of `[1, 2]` is 1.5, so too is the mean of `[date(2023, 1, 1), date(2023, 1, 2)]` now `datetime(2023, 1, 1, 12, 0, 0)`. Let me know if this seems reasonable. If not, it's easy to switch to return dates.
2. The Series `dt` namespace has a defined `mean/median` but the Expr `dt` namespace does not. I don't see why we should have special mean and median functions defined in the `dt` namespace at all, as we should be able to simply call `series.mean()` and it should return the appropriate result. I set this up so that calling `s.mean()` simply dispatches to the `s.dt.mean()`, but I think we should perhaps deprecate `.dt.mean/median`, what do you think?

I also implemented this for the duration types.

### CODE REVIEW

File: crates/polars-core/src/series/implementations/dates_time.rs [hunk 1]
```diff
@@ -172,7 +172,8 @@ macro_rules! impl_dyn_series {
             }
 
             fn mean(&self) -> Option<f64> {
-                self.0.mean()
+                let s = self.0.mean();
+                s
```
Comment (stinodego) [inline crates/polars-core/src/series/implementations/dates_time.rs:176]: This syntax is less readable than the original, I'd recommend reverting this.

**Review by stinodego — CHANGES_REQUESTED** [review]
I'll defer to @MarcoGorelli on this one.

My initial thoughts are: 
* Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.
* I don't really see the benefit of dedicated namespace methods. We can probably deprecate `dt.mean` in favor of `mean`.

File: crates/polars-core/src/series/implementations/dates_time.rs [hunk 2]
```diff
@@ -359,10 +360,10 @@ macro_rules! impl_dyn_series {
                 self.0.min_as_series().$into_logical()
             }
             fn median_as_series(&self) -> Series {
-                Int32Chunked::full_null(self.name(), 1)
-                    .cast(self.dtype())
-                    .unwrap()
-                    .into()
+                Series::new(
+                    self.name(),
+                    &[self.median().map(|v| (v * 86_400_000_000f64) as i64)],
```
Comment (MarcoGorelli) [inline crates/polars-core/src/series/implementations/dates_time.rs:365]: can you use one of the constants here? like `US_IN_DAY`?

File: crates/polars-core/src/series/mod.rs [hunk 3]
```diff
@@ -805,7 +805,13 @@ impl Series {
                 let val = &[self.mean()];
                 Series::new(self.name(), val)
             },
-            dt @ DataType::Duration(_) => {
+            DataType::Date => Series::new(
+                self.name(),
+                &[self.mean().map(|v| (v * 86_400_000_000f64) as i64)],
```
Comment (MarcoGorelli) [inline crates/polars-core/src/series/mod.rs:810]: same

File: py-polars/polars/series/datetime.py [hunk 4]
```diff
@@ -80,15 +83,18 @@ def median(self) -> dt.date | dt.datetime | dt.timedelta | None:
...
         """
         s = wrap_s(self._s)
-        out = s.median()
+        out = s._s.median()
         if out is not None:
             if s.dtype == Date:
-                return _to_python_date(int(out))
-            else:
+                # 86_400_000_000 = microseconds in days
+                return _to_python_datetime(int(out * 86_400_000_000), "us")
```
Comment (MarcoGorelli) [inline py-polars/polars/series/datetime.py:90]: same, is there a constant on the python side?

**Review by MarcoGorelli — COMMENTED** [review]
generally on board with the idea

note that for `interpolate`, `Date` is also preserved:
```python
In [23]: pl.Series([date(2020, 1, 1), None, date(2020, 1, 2)]).interpolate()
Out[23]:
shape: (3,)
Series: '' [date]
[
        2020-01-01
        2020-01-01
        2020-01-02
]
```
, maybe that should also become `Datetime[us, None]`? (that's for a separate PR though)

Also agree with deprecating `.dt.mean` (but, also, that's for a separate PR)

File: crates/polars-core/src/series/implementations/dates_time.rs (hunk 1 above)
Comment (mcrumiller) [reply crates/polars-core/src/series/implementations/dates_time.rs:176]: Ahh yes, I did this during debugging and did not intend to keep this.

**Comment by mcrumiller** [comment]
> I'll defer to @MarcoGorelli on this one.
> 
> My initial thoughts are:
> 
> * Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.
> * I don't really see the benefit of dedicated namespace methods. We can probably deprecate `dt.mean` in favor of `mean`.

@stinodego should I move the date-specific logic into `series.py` then, or should we keep the logic in `datetime.py` and import it, but keep it outside of the `.dt` namespace? In other words, is `datetime.py` *only* reserved for the `dt` namespace, or should it also hold date-specific logic?

**Comment by mcrumiller** [comment]
> * Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.

Both the `mean` and `median` functions map from `i32` or `i64` (date vs datetime) to `f64` to perform the mean anyway. Since `datetime` is backed by `i64` whereas dates are only backed by `i32`, I think there's no issue with date -> datetime conversion: even with the finest time unit of nanoseconds, our maximum datetime is $\frac{2^{63}-1}{1,000,000,000}=9,223,372,036$ days since the unix epoch, compared to `dates` which have a max date of $2^{31}-1=4,294,967,295$ days.

File: py-polars/polars/series/datetime.py (hunk 4 above)
Comment (mcrumiller) [reply py-polars/polars/series/datetime.py]: We have `NS_PER_SECOND` AND `SECONDS_PER_DAY` so I brought those in to create a `US_PER_DAY`. Thanks for the suggestions.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for updating - not sure about the if/then statements in `Series.mean` and co, will take a closer look

File: py-polars/polars/series/series.py [hunk 5]
```diff
@@ -1672,7 +1672,10 @@ def mean(self) -> int | float | None:
         2.0
 
         """
-        return self._s.mean()
+        if self.dtype.is_temporal():
+            return self.dt.mean()
+        else:
+            return self._s.mean()
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply py-polars/polars/series/series.py:1678]: Hye @MarcoGorelli give me a bit, I'm pushing this into the rust side so we can have a cleaner call to `s.mean()` from the python side.

**Comment by mcrumiller** [comment]
@MarcoGorelli I just pushed a new version with the following:
 
* Updated the `mean` and `median` on the rust side so for temporal dtypes it returns the correct dtype directly back to python
* Removed `.dt.mean`, `median`, `min`, and `max` from the `dt` namespace, as these function properly via the normal series functions.
* Removed if/then temporal code from `series.py`.

**Comment by mcrumiller** [comment]
Ahh I should deprecate instead of move.

**Comment by mcrumiller** [comment]
Not sure if the test failure is related.

**Comment by mcrumiller** [comment]
Thanks @stinodego. I went ahead and added `std` for `date`/`datetime` series for issue #12883 as well since it seems to fit in here.

File: py-polars/polars/series/datetime.py [hunk 6]
```diff
@@ -29,87 +33,25 @@ def __getitem__(self, item: int) -> dt.date | dt.datetime | dt.timedelta:
...
-                return _to_python_date(int(out))
-            else:
-                return _to_python_datetime(int(out), s.dtype.time_unit)  # type: ignore[attr-defined]
-        return None
+        """Return maximum as Python datetime."""
+        return self._s.max()
+
+    @deprecate_function("Use `series.median` instead.", version="0.19.20")
+    def median(self) -> dt.datetime | dt.timedelta | None:
+        """Return median as python DateTime."""
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply py-polars/polars/series/datetime.py:48]: Sounds resonable. `pl.Date` is no longer returned but that sounds nitpicky. I also need to add `pl.Time`. Have new PR incoming.

**Comment by MarcoGorelli** [comment]
> Both the `mean` and `median` functions map from `i32` or `i64` (date vs datetime) to `f64` to perform the mean anyway. Since `datetime` is backed by `i64` whereas dates are only backed by `i32`, I think there's no issue with date -> datetime conversion: even with the finest time unit of nanoseconds, our maximum datetime is 263−11,000,000,000=9,223,372,036 days since the unix epoch, compared to `dates` which have a max date of 231−1=4,294,967,295 days.

this calculation is wrong :) for nanoseconds, you need to divide by `1_000_000 * 3600`, which does indeed give a smaller range than `Date` dtype

BUT

in Chrono, both 'us' and 'Date' are limited to dates between +/- 262,000 years since the unix epoch, so in practice, it's not a concern

**Pull request closed without merging**